csharp/src/Apache.Arrow/C/CArrowSchema.cs

```diff
@@ -0,0 +1,584 @@
+        }
+
+        private static IntPtr ConstructDictionary(IArrowType datatype)
+        {
+            if (datatype is DictionaryType dictType)
+            {
+                var cSchema = new CArrowSchema(dictType.ValueType);
+                return cSchema.AllocateAsPtr();
+            }
+            else
+            {
+                return IntPtr.Zero;
+            }
+        }
+
+        /// <summary>
+        /// Initialize the exported C schema as an Arrow type.
+        /// </summary>
+        /// <param name="datatype">The Arrow type to export.</param>
+        public CArrowSchema(IArrowType datatype)
+        {
+            format = StringUtil.ToCStringUtf8(GetFormat(datatype));
+            name = IntPtr.Zero;
+            metadata = IntPtr.Zero;
+            flags = GetFlags(datatype);
+
+            children = ConstructChildren(datatype, out var numChildren);
+            n_children = numChildren;
+
+            dictionary = ConstructDictionary(datatype);
+
+            release = (IntPtr self) =>
+            {
+                var schema = Marshal.PtrToStructure<CArrowSchema>(self);
+
+                Marshal.FreeHGlobal(schema.format);
+                Marshal.FreeHGlobal(schema.name);
+                Marshal.FreeHGlobal(schema.metadata);
+
+                if (schema.n_children > 0)
+                {
+                    for (int i = 0; i < schema.n_children; i++)
+                    {
+                        FreePtr(schema.GetChild(i));
+                    }
+                    Marshal.FreeHGlobal((IntPtr)schema.children);
+                }
+
+                if (schema.dictionary != IntPtr.Zero)
+                {
+                    FreePtr(schema.dictionary);
+                }
+                Marshal.DestroyStructure<CArrowSchema>(self);
+            };
+
+            private_data = IntPtr.Zero;
+        }
+
+        /// <summary>
+        /// Initialize the exported C schema as a field.
+        /// </summary>
+        /// <param name="field">Field to export.</param>
+        public CArrowSchema(Field field) : this(field.DataType)
+        {
+            name = StringUtil.ToCStringUtf8(field.Name);
+            // TODO: field metadata
+            metadata = IntPtr.Zero;
+            flags = GetFlags(field.DataType, field.IsNullable);
+        }
+
+        /// <summary>
+        /// Initialize the exported C schema as a schema.
+        /// </summary>
+        /// <param name="schema">Schema to export.</param>
+        public CArrowSchema(Schema schema) : this(new StructType(schema.Fields.Values.ToList()))
+        {
+            // TODO: top-level metadata
+        }
+
+        /// <summary>
+        /// Allocate an unmanaged pointer and copy this instances data to it.
+        /// </summary>
+        /// <remarks>
+        /// To avoid a memory leak, you must call <see cref="FreePtr"/> on this
+        /// pointer when done using it.
+        /// </remarks>
+        public IntPtr AllocateAsPtr()
```

Comment:
Maybe a nitpick but I think `ExportAsPtr` or maybe just having an override of `Export` would more clearly describe the main intent of this method. I guess `AllocateAndExport` would be more descriptive but is getting a little verbose